crates/matrix-sdk/src/widget/filter.rs

```diff
@@ -103,271 +103,325 @@ pub enum StateEventFilter {
 }
 
 impl StateEventFilter {
-    fn matches(&self, matrix_event: &MatrixEventFilterInput) -> bool {
-        let Some(state_key) = &matrix_event.state_key else {
-            // Message-like event doesn't match a state event filter.
+    fn matches(&self, filter_input: &FilterInput) -> bool {
+        let FilterInput::State(state_filter_input) = filter_input else {
             return false;
         };
 
         match self {
-            StateEventFilter::WithType(event_type) => {
-                matrix_event.event_type == TimelineEventType::from(event_type.clone())
+            StateEventFilter::WithType(filter_type) => {
+                &state_filter_input.event_type == filter_type
             }
             StateEventFilter::WithTypeAndStateKey(event_type, filter_state_key) => {
-                matrix_event.event_type == TimelineEventType::from(event_type.clone())
-                    && state_key == filter_state_key
+                &state_filter_input.event_type == event_type
+                    && state_filter_input.state_key == *filter_state_key
             }
         }
     }
+    fn filter_event_type(&self) -> String {
+        match self {
+            Self::WithType(filter_event_type) => filter_event_type.to_string(),
+            Self::WithTypeAndStateKey(event_type, _) => event_type.to_string(),
+        }
+    }
+}
+
+// Filter input:
+
+/// The input data for the filter. This can either be constructed from a
+/// [`Raw<AnyTimelineEvent>`] or a [`SendEventRequest`].
+#[derive(Debug)]
+pub enum FilterInput {
+    State(FilterInputState),
+    MessageLike(FilterInputMessageLike),
+}
```

Comment:
Considering how these get created it feels like you'd want these to be types that hold a reference to event type instead of taking ownership of the type and content.

Reply:
Where would the owned value of the event type live when we take the reference to it?
If we use the `TryFrom` construction, we dont store the deserialized value anywhere.

Reply:
Very nice.